### ISSUE TITLE
Caching for server side toggles

### ISSUE THREAD

**Issue by Meemaw** [Unleash/unleash-client-nextjs#45]
### Describe the feature request

It seems the current recommended approaches to get toggles with SSR/GSSP both use an un-cached fetch to the API. Since the definitions rarely change, I think it doesn't make much sense to fetch that for every request, but instead keep an unleash client instance around and re-fetch definitions every n seconds.

With this, we can then just do the toggles evaluation for each request.

**Comment by Tymek** [comment]
### App Router
You can do it with App Router server-side fetch cache, example here:
https://github.com/Unleash/unleash-client-nextjs/blob/main/example/src/app/app-page/page.tsx#L22
```ts
  const definitions = await getDefinitions({
    fetchOptions: {
      next: { revalidate: 15 },
    },
  });
```
Docs: https://nextjs.org/docs/app/building-your-application/data-fetching#caching-data

### Pages Router
You can have an [API endpoint that will fetch definitions](https://github.com/Unleash/unleash-client-nextjs/blob/main/example/src/pages/api/proxy-definitions.ts), with [cache-control (Docs)](https://nextjs.org/docs/pages/building-your-application/routing/api-routes#cache-control). Then use this URL as source of definitions. As in [middleware example](https://github.com/Unleash/unleash-client-nextjs/blob/main/example/src/middleware.ts#L17-L25).

**Comment by Tymek** [comment]
# Plans
### Client-side
I'll add an example of how to implement API route "proxying" for frontend API.

**Comment by Meemaw** [comment]
@Tymek thanks for the reply. I'm particularly interested in the `Pages Router`. While cache-control would help with the load on BE, its still a network roundtrip, which will add at least 50ms latency to the SSR render time (assuming its cached on CDN).

I was thinking more of a in-memory cache using something like https://github.com/alexreardon/memoize-one (no need for library, just an example).

**Comment by Tymek** [comment]
You can use caching on SSR. [(docs)](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props#caching-with-server-side-rendering-ssr). Downside is that if you are returning definitions from there to the client, you will expose your configuration. You can filter toggles you're interested in I guess. Will including this in the library help? I'm hesitant because resolving feature toggles on frontend (running server-client on frontend) isn't something Unleash supported before Next.js SDK. 

Correct me if I'm wrong, but I don't think it's possible to store something in-memory in a Serverless/Edge environment. I looked into using [KV store](https://vercel.com/docs/storage/vercel-kv), but definitions are too big for that.

**Comment by Meemaw** [comment]
> You can use caching on SSR. [(docs)](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props#caching-with-server-side-rendering-ssr). Downside is that if you are returning definitions from there to the client, you will expose your configuration. You can filter toggles you're interested in I guess. Will including this in the library help? I'm hesitant because resolving feature toggles on frontend (running server-client on frontend) isn't something Unleash supported before Next.js SDK.
> 
> Correct me if I'm wrong, but I don't think it's possible to store something in-memory in a Serverless/Edge environment. I looked into using [KV store](https://vercel.com/docs/storage/vercel-kv), but definitions are too big for that.

By in memory I don't mean anything fansy, just storing things in a local variable.

**Comment by Tymek** [comment]
I don't think that will run as expected on Cloudflare/Vercel etc. I'll experiment with it later this week. And again, if you have any user-land implementations, sharing it will help :)

**Comment by Meemaw** [comment]
> I don't think that will run as expected on Cloudflare/Vercel etc. I'll experiment with it later this week. And again, if you have any user-land implementations, sharing it will help :)

Well, we are using `unleash-client` for now, but wanted to migrate to `@unleash/nextjs`. We are using a `Unleash` client instance, which will just fetch definitions every n seconds in the background. The client then exposes `unleash.getFeatureToggleDefinitions()` method which just returns these definitions from memory.

**Comment by Tymek** [comment]
Ok. Thank you for explaining 👍🏻 
I'm in favor of wrapping `cache(getDefinitions(...), ttl)` in a functional style instead of OOP instances. It's more flexible. An example use case with long-running Next.js server is definitely something I'd like to cover soon.

**Comment by mltsy** [comment]
I had similar thoughts to @Meemaw when trying to switch to this package for the first time just recently.  I guess the biggest question I have is: why doesn't `@unleash/nextjs` use the Node SDK (`unleash-client`) on the server side?  It seems like that already has all of this stuff worked out, and this Next package could just pass in the applicable environment variables and wrap it up in a nice bow for the user, providing a long-lived server-side client for cases where we're *not* using serverless edge functions (which is the case for me).

I suppose maybe that didn't make as much sense originally, if the original motivation of this package was to support Serverless/Edge, but it's certainly worth supporting considering how simple it ought to be to just wrap up `unleash-client` (and since this package already requires `unleash-client` anyway).

**Comment by Tymek** [comment]
Next.js doesn't use Node SDK, because of incompatibility in handing "Host" and "IP" - deprecated strategies, that cause errors when imported into client-side code. Goal is to have both SDKs unified. I'll check if I can patch Node SDK to work in the Browser.

**Comment by yosiawan** [comment]
@mltsy in your PR (#66) you mentioned that the `flagsClient` is non-syncing does this means we still create multiple instances of Unleash in our app using this method?

```  const flags = flagsClient(toggles); // instantiates a static (non-syncing) unleash-proxy-client```


Anyone know how to maintain a single instance in an SSR setup? It seems my app is causing a timeout in my Unleash server due to having multiple instances

**Comment by mltsy** [comment]
I'm not using unleash anymore, so take this with a grain of salt, but my recollection is that `flagsClient` will create a client, which does not, by default, contact the server.  In the example given, `getDefinitions` is what sends a single request to the server to fetch the flag definitions.  `flagsClient` takes whatever "toggles" you pass in, and stores them in memory for evaluation later.  So if you are creating multiple of them, yes you will have multiple clients, but the unleash server will only be contacted when you call getDefinitions (unless you tell the client to start syncing with the server - I don't recall how that is done)
